src/Gravity_impl.hpp

```diff
@@ -0,0 +1,652 @@
+			grad_phi_curr[level][n]->setVal(0.);
+		}
+	}
+
+	finest_level_allocated = level;
+}
+
+template <typename T> auto Gravity<T>::get_gravity_type() -> gravity::GravityMode { return gravity::gravity_type; }
+
+template <typename T> auto Gravity<T>::get_max_solve_level() -> int { return gravity::max_solve_level; }
+
+template <typename T> auto Gravity<T>::NoSync() -> int { return gravity::no_sync; }
+
+template <typename T> auto Gravity<T>::DoCompositeCorrection() -> int { return gravity::do_composite_phi_correction; }
+
+// if true, check the residuals manually
+template <typename T> auto Gravity<T>::test_results_of_solves() -> int { return test_solves; }
+
+template <typename T> auto Gravity<T>::get_grad_phi_prev(int level) -> amrex::Vector<std::unique_ptr<MultiFab>> & { return grad_phi_prev[level]; }
+
+template <typename T> auto Gravity<T>::get_grad_phi_prev_comp(int level, int comp) -> MultiFab * { return grad_phi_prev[level][comp].get(); }
+
+template <typename T> auto Gravity<T>::get_grad_phi_curr(int level) -> amrex::Vector<std::unique_ptr<MultiFab>> & { return grad_phi_curr[level]; }
+
+template <typename T> void Gravity<T>::plus_grad_phi_curr(int level, amrex::Vector<std::unique_ptr<MultiFab>> &addend)
+{
+	for (int n = 0; n < AMREX_SPACEDIM; n++) {
+		grad_phi_curr[level][n]->plus(*addend[n], 0, 1, 0);
+	}
+}
+
+template <typename T> void Gravity<T>::swapTimeLevels(int level)
+{
+	BL_PROFILE("Gravity<T>::swapTimeLevels()");
+
+	if (gravity::gravity_type == GravityMode::Poisson) {
+		for (int n = 0; n < AMREX_SPACEDIM; n++) {
+			std::swap(grad_phi_prev[level][n], grad_phi_curr[level][n]);
+			grad_phi_curr[level][n]->setVal(1.e50);
+		}
+	}
+}
+
+template <typename T> void Gravity<T>::solve_for_phi(int level, MultiFab &phi, const amrex::Vector<MultiFab *> &grad_phi, int is_new)
+{
+	BL_PROFILE("Gravity<T>::solve_for_phi()");
+
+	if (gravity::verbose > 1) {
+		amrex::Print() << " ... solve for phi at level " << level << std::endl;
+	}
+
+	const Real strt = amrex::ParallelDescriptor::second();
+
+	if (is_new == 0) {
+		sanity_check(level);
+	}
+
+	Real time = NAN;
+	if (is_new == 1) {
+		time = sim->tNew_[level];
+	} else {
+		time = sim->tOld_[level];
+	}
+
+	// If we are below the max_solve_level, do the Poisson solve.
+	// Otherwise, interpolate using a fillpatch from max_solve_level.
+
+	if (level <= gravity::max_solve_level) {
+
+		amrex::Vector<amrex::MultiFab *> phi_p(1, &phi);
+
+		const auto &rhs = get_rhs(level, 1, is_new);
+
+		amrex::Vector<amrex::Vector<amrex::MultiFab *>> grad_phi_p(1);
+		grad_phi_p[0].resize(AMREX_SPACEDIM);
+		for (int i = 0; i < AMREX_SPACEDIM; i++) {
+			grad_phi_p[0][i] = grad_phi[i];
+		}
+
+		amrex::Vector<amrex::MultiFab *> res_null;
+
+		level_solver_resnorm[level] = solve_phi_with_mlmg(level, level, phi_p, amrex::GetVecOfPtrs(rhs), grad_phi_p, res_null, time);
+
+	} else {
+		sim->FillCoarsePatch(level, time, phi, phi_old_, phi_new_, BCs_cc_, quokka::centering::cc, quokka::direction::na);
```

Comment:
@BenWibking in this line, you only had 5 arguments (the first five) when you wrote this file 2 years ago. Since then, I think the code has been modified so the function `FillCoarsePatch` needs 8 arguments. I added three more arguments based on how this function is called in `simulation.hpp`. But now I get the error:

In file included from /Users/psharda/Documents/quokka/src/Gravity.hpp:488:
/Users/psharda/Documents/quokka/src/Gravity_impl.hpp:264:62: error: use of undeclared identifier 'BCs_cc_'
                sim->FillCoarsePatch(level, time, phi, phi_old_, phi_new_, BCs_cc_, quokka::centering::cc, quokka::direction::na);

Reply:
I dont know if this thing here has to be face centered or cell centered. I think a direction has to be specified if it is face centered...

Reply:
`phi` is always cell centered.

Reply:
Okay, so this should be fine. How do I get `BCs_cc_` to be known to this function

Reply:
`sim` is a pointer to the `AMRSimulation` class, so you can access it with `sim->BCs_cc_`.

Reply:
```
[ 26%] Building CXX object src/StarCluster/CMakeFiles/star_cluster.dir/star_cluster.cpp.o
In file included from /Users/psharda/Documents/quokka/src/StarCluster/star_cluster.cpp:27:
In file included from /Users/psharda/Documents/quokka/src/RadhydroSimulation.hpp:60:
In file included from /Users/psharda/Documents/quokka/src/Gravity.hpp:488:
/Users/psharda/Documents/quokka/src/Gravity_impl.hpp:264:67: error: 'BCs_cc_' is a protected member of 'AMRSimulation<StarCluster>'
                sim->FillCoarsePatch(level, time, phi, phi_old_, phi_new_, sim->BCs_cc_, quokka::centering::cc, quokka::direction::na);
```

Reply:
Ah, right. You can make it a member variable of the `Gravity` class, and then pass it to that class when creating it.

Reply:
how